docs/source/extending/WritingTIAndContextProviders.rst

```diff
@@ -0,0 +1,460 @@
+
+The value of each item in the queries dictionary should be an
+instance of an ``APILookupParams`` class or one derived from it.
+
+.. note:: APILookupParams is an attrs class, if you create a subclass
+    from it you should also make that an attrs class.
+
+Several of the values of this class can have substitutable parameters
+where runtime values (e.g. the observable value)
+are inserted before making the request.
+
+**path**
+
+The sub-path for the query for this IoC type. This will be
+appended to the ``_BASE_URL`` value.
+
+.. code:: python3
+
+    _QUERIES = _QUERIES = {
+        "ipv4": APILookupParams(path="/api/v1/indicators/IPv4/{observable}/general"),
+
+In this example you can see that "{observable}", the IoC value, is
+a substitutable parameter.
+
+**verb**
+
+The provider framework currently only supports "GET"
+
+**full_url**
+If True, the ``_BASE_URL`` value is ignored and the ``path`` value
+is treated as a full URL and used in the request as-is.
+
+**headers**
+
+A dictionary of request headers.
+This also supports parameter substitution of any value surrounded with
+braces.
+
+Example:
+
+.. code:: python3
+
+     _QUERIES = _QUERIES = {
+        "ipv4": APILookupParams(
+            path="/api/v1/indicators/IPv4/{observable}/general"
+            headers = {"X-OTX-API-KEY": "{AuthKey}"}
+        ),
+        ...
+
+**params**
+A dictionary of request parameters. This also supports
+parameter substitution of values:
+
+.. code:: python3
+
+     _QUERIES = _QUERIES = {
+        "ipv4": APILookupParams(
+            path="/api/v1/indicators/IPv4"
+            params={"iocValue": "{observable}"},
+        ),
+        ...
+
+**data**
+This is currently not supported but we will implement if and
+when required. This is a dictionary that will be supplied as request data. supports
+parameter substitution for values.
+
+**auth_type**
+Currently only "HTTPBasic" is supported.
+
+**auth_str**
+This is an list of values to supply as the request ``auth`` property.
+Supports substitution.
+
+.. code:: python3
+
+     _QUERIES = _QUERIES = {
+        "ipv4": APILookupParams(
+            path="/api/v1/indicators/IPv4"
+            auth_str = ["{ApiID}", "{AuthKey}"],
+        ),
+        ...
+
+**sub_type**
+Not currently used.
+
+The parse_results method
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+See :py:meth:`parse_results <msticpy.context.tiproviders.ti_provider_base.parse_results>`
+for the method header.
+
+This method is responsible for taking the JSON response (as
+a Python dictionary) and extracting and returning severity
+and relevant details.
+
+The implementation in the example at the start of this document
+(and below) shows a simple process, but it can be as complex as needed.
+
+.. note:: we would recommend creating child methods to handle
+    different response types if you need to do complex
+    processing.
+
+.. code:: python3
+
+    def parse_results(self, response: Dict) -> Tuple[bool, ResultSeverity, Any]:
```

Comment:
Is `Tuple[bool, ResultSeverity, Any]` the required return type or just what you've chosen for this example? If the former, then creating/re-using a new type would be easier than copy/pasting that annotation for each `parse_results` method.

Reply:
Great idea! I normally now do that as a rule (returning raw tuples is nasty so I usually use named tuples), but this a hangover of prehistoric code from the early days of msticpy.
I'll implement that.